browser/brave_cosmetic_resources_tab_helper.cc

```diff
@@ -0,0 +1,499 @@
+        pre_init_script += "true";
+      } else {
+        pre_init_script += "false";
+      }
+      pre_init_script += ";"
+        "}"
+        "if (window.generichide === undefined) {"
+          "window.generichide = ";
+      if (generichide) {
+        pre_init_script += "true";
+      } else {
+        pre_init_script += "false";
+      }
+      pre_init_script += ";"
+        "}"
+        "})();";
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(pre_init_script),
+        base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+        base::UTF8ToUTF16(*BraveCosmeticResourcesTabHelper::observing_script_),
+        base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+    if (to_inject.length() > 1) {
+      std::string script =
+        "(function() {"
+          "let script;"
+          "let text = " + to_inject + ";"
+          "try {"
+            "script = document.createElement('script');"
+            "const textnode = document.createTextNode(text);"
+            "script.appendChild(textnode);"
+            "(document.head || document.documentElement).appendChild(script);"
+          "} catch (ex) {"
+          "}"
+          "if (script) {"
+            "if (script.parentNode) {"
+              "script.parentNode.removeChild(script);"
+            "}"
+            "script.textContent = '';"
+          "}"
+        "})();";
+      frame_host->ExecuteJavaScriptInIsolatedWorld(
+          base::UTF8ToUTF16(script),
+          base::NullCallback(), ISOLATED_WORLD_ID_CHROME_INTERNAL);
+    }
+    // Working on css rules, we do that on a main frame only
+    if (!do_non_scriplets)
+      return;
+    CSSRulesRoutine(url, resources_dict, frame_id);
+  }
+}
+
+void BraveCosmeticResourcesTabHelper::CSSRulesRoutine(
+    const std::string& url_string, base::DictionaryValue* resources_dict,
+    content::GlobalFrameRoutingId frame_id) {
+  // Check are first party cosmetic filters enabled
+  const GURL url(url_string);
+  if (url.is_empty() || !url.is_valid() || IsVettedSearchEngine(url.host())) {
+    return;
+  }
+
+  base::ListValue* cf_exceptions_list;
+  if (resources_dict->GetList("exceptions", &cf_exceptions_list)) {
+    for (size_t i = 0; i < cf_exceptions_list->GetSize(); i++) {
+      exceptions_.push_back(cf_exceptions_list->GetList()[i].GetString());
+    }
+  }
+  base::ListValue* hide_selectors_list;
+  base::ListValue* force_hide_selectors_list = nullptr;
+  if (resources_dict->GetList("hide_selectors", &hide_selectors_list)) {
+    if (enabled_1st_party_cf_filtering_) {
+      force_hide_selectors_list = hide_selectors_list;
+    } else {
+      std::string cosmeticFilterConsiderNewSelectors_script =
+          "(function() {"
+            "let nextIndex = window.cosmeticStyleSheet.rules.length;";
+      std::string json_selectors;
+      base::JSONWriter::Write(*hide_selectors_list, &json_selectors);
```

Comment:
should probably throw an error or do https://github.com/brave/brave-core/pull/6735/files#diff-cfde9d579db91c98f9da4fe25bdea104c9d925ac00346fe153e9d8efa9fffc97R294 if `base::JSONWriter::Write` does not return true.

Reply:
it's a `json_selectors` length validation on empty and it does `const selectors = []` if it's empty, but ok, I can include check on false as well, not a big deal

Reply:
addressed